src/Core/Waiter.php

```diff
@@ -0,0 +1,222 @@
+<?php
+
+declare(strict_types=1);
+
+namespace AsyncAws\Core;
+
+use AsyncAws\Core\Exception\Http\HttpException;
+use AsyncAws\Core\Exception\Http\NetworkException;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * The result promise is always returned from every API call. Remember to call `resolve()` to
+ * make sure the request is actually sent.
+ */
+class Waiter
+{
+    public const STATE_SUCCESS = 'success';
+    public const STATE_FAILURE = 'failure';
+    public const STATE_PENDING = 'pending';
+
+    protected const WAIT_TIMEOUT = 30.0;
+    protected const WAIT_DELAY = 5.0;
+
+    /**
+     * @var AbstractApi|null
+     */
+    protected $awsClient;
+
+    /**
+     * @var array|object|null
+     */
+    protected $request;
```

Comment:
Only a minor thing. Rename this to “input” and add some description. 

Or maybe I’m just confused, but I thought this was a request object...

Reply:
It will never be an array BTW.

Reply:
Fixed. Renamed Result's request too for consistency